### PR TITLE
Build an NPM package

### PR DESCRIPTION
**Pull request by alessiostalla**
This covers partially issue #9.
I've tested the built package in a Node project. I haven't checked browser compatibility at all.

### CODE REVIEW

**Comment by alessiostalla** [comment]
OK, now it works, at least locally. I had to downgrade `nanoid` to the latest 3.x version that, as far as I understand, is maintained alongside 4.x with the same features, but supports CommonJS whereas 4.x only supports ESM (and ESM support on Node is experimental and turned off by default).

File: deno.lock [hunk 1]
```diff
@@ -283,13 +296,18 @@
   "npm": {
     "specifiers": {
       "littoral-templates@0.2.2": "littoral-templates@0.2.2",
+      "nanoid@3.3.4": "nanoid@3.3.4",
```
Comment (dslmeinte) [inline deno.lock:299]: I would expect that version 4.0.0 is really _replaced_ by 3.3.4, but like this it's added. (Also below.)

File: scripts/build_npm.ts [hunk 2]
```diff
@@ -1,6 +1,23 @@
...
+await emptyDir("./npm/script/src/m3/ecore/test");
+await emptyDir("./npm/esm/src/m3/test");
+await emptyDir("./npm/script/src/m3/test");
+
+await copy("diagrams", "npm/esm/diagrams", { overwrite: true });
+await copy("diagrams", "npm/script/diagrams", { overwrite: true });
+await copy("models", "npm/esm/models", { overwrite: true });
+await copy("models", "npm/script/models", { overwrite: true });
+await copy("schemas", "npm/esm/schemas", { overwrite: true });
+await copy("schemas", "npm/script/schemas", { overwrite: true });
```
Comment (dslmeinte) [inline scripts/build_npm.ts:16]: Does this mean that we have now 2 NPM packages: one for ESM and one for CommonJS?

File: scripts/build_npm.ts [hunk 3]
```diff
@@ -1,6 +1,23 @@
...
+await copy("src/m3/ecore/test/library.ecore", "npm/esm/src/m3/ecore/test/library.ecore", { overwrite: true });
```
Comment (dslmeinte) [inline scripts/build_npm.ts:17]: I think all test-artifacts shouldn't end up in `npm/`.

File: src/m3/test/json-validator.ts [hunk 4]
```diff
@@ -25,7 +25,7 @@ export const createJsonValidatorForSchema = (schema: unknown): (json: unknown) =
 }
 
 
-const metaSchema = JSON.parse(Deno.readTextFileSync(`${__dirname(import.meta)}/json.schema.json`))
+const metaSchema = JSON.parse(Deno.readTextFileSync(`src/m3/test/json.schema.json`))
```
Comment (dslmeinte) [inline src/m3/test/json-validator.ts:28]: `__dirname(import.meta)` wasn't working on the Node.js-side?

File: deno.lock (hunk 1 above)
Comment (alessiostalla) [reply deno.lock:299]: Right. I didn't edit the lock file by hand. Maybe there's some other code depending on 4.0.0 that didn't trigger the issue.

File: scripts/build_npm.ts (hunk 2 above)
Comment (alessiostalla) [reply scripts/build_npm.ts:16]: Yes. The CommonJS one is for Node and the ESM one is, I presume, for the browser (but I didn't test it in the browser and I would be surprised if it worked as-is). That's how [Deno's dnt tool](https://github.com/denoland/dnt) works.

File: scripts/build_npm.ts (hunk 3 above)
Comment (alessiostalla) [reply scripts/build_npm.ts:17]: I would think so, too. But I don't know Deno's conventions, and seeing that they are part of the source and the "dnt" tool explicitly runs them and finds them where they are, I assumed that's standard practice for Deno packages. I think we can safely remove them after the build step.

File: deno.lock (hunk 1 above)
Comment (dslmeinte) [reply deno.lock:299]: Running `deno task lock-deps` on the commandline should fix that.

File: src/m3/test/json-validator.ts (hunk 4 above)
Comment (alessiostalla) [reply src/m3/test/json-validator.ts:28]: Nope. It errors with
```
Running tests in ./script/m3/test/library.test.js...

/Users/alessio/projects/lioncore-typescript/npm/script/m3/test/json-validator.js:50
const metaSchema = JSON.parse(dntShim.Deno.readTextFileSync(`${(0, local_files_js_1.__dirname)(import.meta)}/json.schema.json`));
                                                                                                      ^^^^

SyntaxError: Cannot use 'import.meta' outside a module
    at Object.compileFunction (node:vm:352:18)
    at wrapSafe (node:internal/modules/cjs/loader:1031:15)
    at Module._compile (node:internal/modules/cjs/loader:1065:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
```

etc. etc.

**Review by dslmeinte — APPROVED** [review]
Let's merge as-is, and tweak later.

File: scripts/build_npm.ts (hunk 3 above)
Comment (dslmeinte) [reply scripts/build_npm.ts:17]: OK, we'll look how it works exactly after merging ;)

**Comment by ftomassetti** [comment]
Hurray!

**Pull request merged**